node/subsystem-util/src/reputation.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A utility abstraction to collect and send reputation changes.
+
+use polkadot_node_network_protocol::{self as net_protocol, PeerId, UnifiedReputationChange};
+use polkadot_node_subsystem::{messages::NetworkBridgeTxMessage, overseer};
+use std::time::Duration;
+
+/// Default delay for sending reputation changes
+pub const REPUTATION_CHANGE_INTERVAL: Duration = Duration::from_secs(30);
+
+/// Collects reputation changes and sends them in one batch to relieve network channels
+#[derive(Debug, Clone)]
+pub struct ReputationAggregator {
+	send_immediately_if: fn(UnifiedReputationChange) -> bool,
+	by_peer: std::collections::HashMap<PeerId, i32>,
+}
+
+impl Default for ReputationAggregator {
+	fn default() -> Self {
+		Self::new(|rep| matches!(rep, UnifiedReputationChange::Malicious(_)))
+	}
+}
+
+impl ReputationAggregator {
+	/// New ReputationAggregator
+	///
+	/// # Arguments
+	///
+	/// * `send_immediately_if` - A function, takes `UnifiedReputationChange`,
+	/// results shows if we need to send the changes right away.
+	/// Useful for malicious changes and tests.
+	pub fn new(send_immediately_if: fn(UnifiedReputationChange) -> bool) -> Self {
+		Self { by_peer: Default::default(), send_immediately_if }
+	}
+
+	/// Sends collected reputation changes in a batch,
+	/// removing them from inner state
+	pub async fn send(
+		&mut self,
+		sender: &mut impl overseer::SubsystemSender<NetworkBridgeTxMessage>,
+	) {
+		for (&peer_id, &score) in &self.by_peer {
+			sender
+				.send_message(NetworkBridgeTxMessage::ReportPeer(
+					peer_id,
+					net_protocol::ReputationChange::new(score, "Aggregated reputation change"),
+				))
+				.await;
+		}
+		self.by_peer.clear();
+	}
+
+	/// Adds reputation change to inner state,
+	/// сhecks if the change is dangerous, sends all collected changes in a batch if it is
```

Comment:
Looks reasonable,but then we're going to send hashmaps instead of a single message so we will need to prepare a hashmap only for a malicious change. Is it ok?

Reply:
We can keep current message for the malicious ones and add another to send them in bull